kernel-tests: Add test to diff dmesg log

chaitu236 opened the pull request:
Add a new test that diffs the dmesg log with a previous version obtained from a database and uploads the latest log to this database.

### Testing
- [x] Built `kernel-tests` recipe, ensured `kernel-tests-ptest.ipk` is installable
- [x] Ran with current and next kernels on a VM.
- [x] Ran on actual h/w with the db populated with old logs.

chaitu236 committed:
kernel-tests: Add test to diff dmesg log

Add a new test that diffs the dmesg log with a previous version obtained
from a database and uploads the latest log to this database.

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

bstreiff commented on recipes-kernel/kernel-tests/kernel-tests.bb:
```diff
@@ -2,7 +2,7 @@ SUMMARY = "Linux kernel-specific tests"
 SECTION = "tests"
 LICENSE = "GPLv2 & GPLv2+"
-LIC_FILES_CHKSUM = "file://run-ptest;md5=0c0d48c031233bf837da6482d67473ae"
+LIC_FILES_CHKSUM = "file://run-ptest;md5=7c0a0ce46f137996a104bdfa677dbb44"
```
This doesn't seem correct; the `run-ptest` file doesn't contain license text at all.

Should this just be `"file://${COMMON_LICENSE_DIR}/GPL-2.0-or-later;md5=fed54355545ffd980b814dab4a3b312c`?

(This is probably logically a separate commit.)

chaitu236 replied:
Fixed

amstewart commented on recipes-kernel/kernel-tests/kernel-tests.bb:
```diff
@@ -82,4 +86,8 @@ do_install_ptest:append:xilinx-zynqhf() {
 }
 
+pkg_postinst_ontarget:${PN}-ptest:append() {
+    python3 -m pip install pymongo
```
Why aren't we just distributing mongodb from the distro feeds? Licensing?

chaitu236 replied:
I hadn't realized that the feeds already contained `python3-pymongo`. Updated the PR to use it instead.

chaitu236 committed:
kernel-tests: Fix license file

kernel-performance-tests and kernel-tests recipes' LIC_FILES_CHKSUM
points to files that don't have any license. So replace them with the
generic GPL-2.0-or-later file.

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

gratian commented on recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py:
```diff
@@ -0,0 +1,157 @@
+    if not results:
+        return None
+
+    for result in results:
```
This for loop reads weird to me since we're returning with the first 'result' in 'results' i.e. do we need the loop?

chaitu236 replied:
`results` is a generator and if we don't want to use the loop method can use `next(results)` but this fails if there are no results - so it would require putting a try catch block which ends up complicating the logic even more. This seemed the simplest option to me.

But if someone knows a better way, I'm happy to change this.

gratian replied:
OK, I think it's fine.

chaitu236 replied:
Modified the implementation to account for the new 'prefer same major.minor' logic and got rid of this loop.

gratian commented on recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py:
```diff
@@ -0,0 +1,157 @@
+
+def diff_logs(current_log, old_log):
+    current_log = prepare_log_for_diff(current_log)
+    old_log = prepare_log_for_diff(old_log)
```
Is the test going to fail if we don't have a 'old_log' to compare against? Maybe that's OK for the first test run.

chaitu236 replied:
Yes, if there isn't an old log, `old_log` is empty string and the test will fail as diff will be non-empty.

I'm planning to seed the db with 22.5 BSI's kernel log. But the test will likely fail - at least the first time anyway as there are so many changes and it is not possible to put all those in allow list.

gratian commented on recipes-kernel/kernel-tests/kernel-tests-files/kernel_dmesg_diff.py:
```diff
@@ -0,0 +1,157 @@
+def get_old_dmesg_log(db):
+    query = {}
+    query['kernel_version'] = {'$ne': get_kernel_version()}
+    query['kernel_type'] = 'next' if is_next_kernel() else 'current'
```
In theory you should be able to compare against logs from '-next' kernels too. For example when we upgrade the current 5.15 kernel to the 'next' 6.0 one we should be able to use the '-next' logs already stored for 6.0 because of '-next' testing.

I'm also wondering if the query should be restricted to logs from the same 'major.minor' set as the current kernel. We might get failures just because dmesg output changes between major kernel versions.

chaitu236 replied:
I'll try to make this more intelligent and pick existing `major.minor` log if it isn't too involved.

Don't the `major.minor` versions change every couple of months? I think it would still be valuable to diff against any old log if the one with same `major.minor` isn't available - the test will fail but we can still manually vet the diff from the test output.

gratian replied:
I see your point  - for 'next' it will change every couple of months, for the stable shipping kernel is less often but yes it would still be valuable to see the diff.

chaitu236 replied:
Updated to prefer last log with same kernel major.minor version, but if not then it'll pick last log of same kernel type (current/next).
